pythonforandroid/recipes/python2/patches/fix-api-minor-than-21.patch

```diff
@@ -0,0 +1,145 @@
++++ Python-2.7.15/configure.ac	2018-07-05 17:44:50.500985727 +0200
+@@ -1790,7 +1790,7 @@ fi
+ # structures (such as rlimit64) without declaring them. As a
+ # work-around, disable LFS on such configurations
+ 
+-use_lfs=yes
++use_lfs=no
+ AC_MSG_CHECKING(Solaris LFS bug)
+ AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[
+ #define _LARGEFILE_SOURCE 1
+diff -Naurp Python-2.7.15.orig/Modules/_localemodule.c Python-2.7.15/Modules/_localemodule.c
+--- Python-2.7.15.orig/Modules/_localemodule.c	2018-04-30 00:47:33.000000000 +0200
++++ Python-2.7.15/Modules/_localemodule.c	2018-07-05 16:39:08.241023323 +0200
+@@ -170,6 +170,12 @@ PyLocale_setlocale(PyObject* self, PyObj
+         PyErr_SetString(Error, "invalid locale category");
+         return NULL;
+     }
++#else
++    #ifdef __ANDROID__
++        #if defined(__ANDROID_API__) && __ANDROID_API__ < 20
++            return PyUnicode_FromFormat("%s", "C");
++        #endif
++    #endif
+ #endif
+ 
+     if (locale) {
+@@ -215,7 +221,15 @@ PyLocale_localeconv(PyObject* self)
+         return NULL;
+ 
+     /* if LC_NUMERIC is different in the C library, use saved value */
+-    l = localeconv();
++    //PMPP API<21
++    #if __ANDROID_API__ < 21
++        /* Don't even try on Android's broken locale.h. */
++        goto failed;
++    #else
++         /* if LC_NUMERIC is different in the C library, use saved value */
++         l = localeconv(); //PATCHED
++    #endif
++    //PMPP API<21
+ 
+     /* hopefully, the localeconv result survives the C library calls
+        involved herein */
+diff -Naurp Python-2.7.15.orig/Modules/mmapmodule.c Python-2.7.15/Modules/mmapmodule.c
+--- Python-2.7.15.orig/Modules/mmapmodule.c	2018-04-30 00:47:33.000000000 +0200
++++ Python-2.7.15/Modules/mmapmodule.c	2018-07-05 16:18:40.953035027 +0200
+@@ -78,6 +78,12 @@ my_getpagesize(void)
+ #  define MAP_ANONYMOUS MAP_ANON
+ #endif
+ 
++//PMPP API<21
++#if __ANDROID_API__ < 21
++    extern void *mmap(void *addr, size_t length, int prot, int flags, int fd, off_t offset);
++#endif
++//PMPP API<21
++
+ static PyObject *mmap_module_error;
+ 
+ typedef enum
+diff -Naurp Python-2.7.15.orig/Modules/posixmodule.c Python-2.7.15/Modules/posixmodule.c
+--- Python-2.7.15.orig/Modules/posixmodule.c	2018-04-30 00:47:33.000000000 +0200
++++ Python-2.7.15/Modules/posixmodule.c	2018-07-05 16:20:48.933033807 +0200
+@@ -9477,6 +9477,12 @@ all_ins(PyObject *d)
+ #define MODNAME "posix"
+ #endif
+ 
++//PMPP API<21
++#if __ANDROID_API__ < 21
++    extern ssize_t sendfile(int out_fd, int in_fd, off_t *offset, size_t count);
++#endif
++//PMPP API<21
++
+ PyMODINIT_FUNC
+ INITFUNC(void)
+ {
+diff -Naurp Python-2.7.15.orig/Modules/signalmodule.c Python-2.7.15/Modules/signalmodule.c
+--- Python-2.7.15.orig/Modules/signalmodule.c	2018-04-30 00:47:33.000000000 +0200
++++ Python-2.7.15/Modules/signalmodule.c	2018-07-05 16:40:46.601022385 +0200
+@@ -32,6 +32,13 @@
+ #include <process.h>
+ #endif
+ 
++//PMPP API<21
++#if __ANDROID_API__ < 21
```

Comment:
shouldn't this be `  +#if defined(__ANDROID_API__) && __ANDROID_API__ < 21` instead? to make it more mergeable for upstream

Reply:
Also I'm thinking, maybe the entire lower-than-21 patch should just be dropped. If you look here: https://developer.android.com/about/dashboards/ pre-5.0 phones are now at less than 25% and this problem will kinda resolve itself soonish, so maybe we just shouldn't support building with lower than NDK API 21 anymore.

Reply:
> shouldn't this be ` +#if defined(__ANDROID_API__) && __ANDROID_API__ < 21` instead?

Yes
Note: this patch has not been tested (is disabled and I put in there in the case that we will need it)

Reply:
Update: I have tested the mentioned patch, building with minimum api 19 and, even that the patch successfully applies, the python2 build will fail at linking stage complaining about missing  localeconv...As far as I know localeconv is not fully implemented in android until sdk 21 (https://stackoverflow.com/questions/44736135/ndk-clang-error-undefined-reference-to-localeconv) so I'm afraid that we will not be able to run with a minimum api lower than 21, like the python3 or numpy (#1450) recipes...unless someone find some solution, which I searched but not found :<...for now I leave the patch but maybe we should remove it as proposed by @JonasT...